grails-app/services/org/pih/warehouse/invoice/InvoiceService.groovy

```diff
@@ -152,6 +153,18 @@ class InvoiceService {
         invoiceItem.delete()
     }
 
+    def addItems(Invoice invoice, List items) {
+        items.each { item ->
+            InvoiceCandidate candidateItem = InvoiceCandidate.get(item.id)
```

Comment:
It feels like an additional query here seems unnecessary. Why are we doing this when we could be using the POST data?

Reply:
In post we send only candidate id and quantityToInvoice. I could send all data back, but then I will need to find BudgetCode by code, GlAccount by code and uom basing on the formatted string value (`"${quantityUom?.code}/${quantityPerUom as Integer}"` made in transient). I feel the current version is smoother and cleaner

Reply:
Ok sounds good. One issue we might need to deal with is when the candidate invoice item does not exist any longer (i.e. it was added to another invoice in the time between it was queried and saved in our invoice). Right now it seems like we're just silently ignoring that case but that could become confusing. It shouldn't happen very often but it'll be annoying to the user when it does.

Reply:
So let's throw an IllegalArgumentException or ValidationException for that case.

Reply:
Ok, I'll do it as a first thing tomorrow.